git-server/src/lib.rs

```diff
@@ -305,54 +318,68 @@ pub async fn run(options: Options) -> anyhow::Result<()> {
-        .and(warp::filters::addr::remote())
-        .and(path::param())
-        .and(peer_id)
-        .and(path::peek())
-        .and(
-            warp::filters::query::raw()
-                .or(warp::any().map(String::default))
-                .unify(),
+    let app = git_handler
+        .layer(Extension(ctx.clone()))
+        .layer(
+            TraceLayer::new_for_http()
+                .make_span_with(DefaultMakeSpan::new().level(Level::INFO))
+                .on_response(
+                    |response: &Response<BoxBody>, latency: Duration, _span: &Span| {
+                        tracing::info!("STATUS={:?} LATENCY={:?}", response.status(), latency)
+                    },
+                ),
         )
-        .and_then(git_handler)
-        .recover(recover)
-        .with(warp::log("radicle_git_server"));
-    let server = warp::serve(server);
+        .into_make_service_with_connect_info::<SocketAddr>();
 
     if let (Some(cert), Some(key)) = (options.tls_cert, options.tls_key) {
-        server
-            .tls()
-            .cert_path(cert)
-            .key_path(key)
-            .run(options.listen)
-            .await;
+        let config = RustlsConfig::from_pem_file(cert, key).await.unwrap();
+
+        tracing::info!("listening on https://{}", options.listen);
+        axum_server::bind_rustls(options.listen, config)
+            .serve(app)
+            .await
+            .unwrap();
     } else {
-        server.run(options.listen).await;
+        tracing::info!("listening on http://{}", options.listen);
+        axum::Server::bind(&options.listen)
+            .serve(app)
+            .await
+            .unwrap();
     }
+
     Ok(())
 }
 
 async fn git_handler(
-    ctx: Context,
+    Extension(ctx): Extension<Context>,
     method: Method,
     headers: HeaderMap,
-    body: impl Buf,
-    remote: Option<net::SocketAddr>,
-    project_id: String,
-    peer_id: Option<PeerId>,
-    request: warp::filters::path::Peek,
-    query: String,
-) -> Result<Box<dyn Reply>, Rejection> {
-    let remote = remote.expect("There is always a remote for HTTP connections");
+    body: Bytes,
+    ConnectInfo(remote): ConnectInfo<SocketAddr>,
+    uri: Uri,
+    query: RawQuery,
+) -> impl IntoResponse {
+    let query = query.0.unwrap_or_else(|| "".to_string());
+    let (peer_id, project_id, request) = {
+        let tokens = uri.path().split('/').collect::<Vec<&str>>();
+        let peer_id: Option<PeerId> = tokens.first().map(|p| p.parse().ok()).unwrap();
+        let (project_id, request) = if peer_id.is_none() {
+            if tokens.len() < 3 {
+                return Err(Error::ServiceUnavailable(""));
+            }
+
+            (tokens[1], tokens[2..].join("/"))
+        } else {
+            if tokens.len() < 4 {
+                return Err(Error::ServiceUnavailable(""));
+            }
+
+            (tokens[2], tokens[3..].join("/"))
+        };
+
+        (peer_id, project_id, request)
+    };
```

Comment:
Not a fan of manually parsing the URL path... there must be a better way, or perhaps for now drop support for optional PeerId, as we aren't using it yet.

Reply:
`PeerId` is not the issue here per se. I'm using [`MakeService`](https://github.com/radicle-dev/radicle-client-services/blob/221ab2a75a7b7774964a64eded3f510cca480713/git-server/src/lib.rs#L332) (which was suggested in axum channel) since we don't need a `Router`.

If you find this problematic, I can use `any` in combination with `Router` but that has performance cost.

Personally, I'd rather write tests for extraction logic than take performance hit, but I can rewrite it to take `Router`/`any` if you want.

Reply:
IMO it's the job of the router/framework to parse URLs, not us. I'd either drop the peer-id parsing or take the performance hit (which I doubt is even measurable).

Reply:
https://github.com/radicle-dev/radicle-client-services/compare/479c88c6c5672ee135534cf98191b991e0d12c77..cd2a351c51c38d58db8473998e0aefdfe1f168db

Reply:
`PeerId` logic is in there, so we can add that back in, in future.